In [4]:
import pandas as pd
from sklearn.metrics import confusion_matrix

### 1.Create a new file named model_evaluation.py or model_evaluation.ipynb for these exercises.`


1. Get the actual results
2. Obtain model predictions
3. decide positive case 
4. Produce a confusion matrix
5. Produce our metrics
positice : cat prediction
false positive:  predicting a cat when we actually have a dog (7)
false negative: predicting a dog, when we actually have a cat. (13)
False Negative : a miss.
False positive : a false alarm.

### 2. Given the following confusion matrix, evaluate (by hand) the model's performance.`


|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |

- In the context of this problem, what is a false positive?

false positive:  predicting a cat when we actually have a dog (7)

- In the context of this problem, what is a false negative?

false negative: predicting a dog, when we actually have a cat. (13)

- How would you describe this model?

In [6]:

# reality is either true/false
# prediction is either positive/negative
true_positives = 34
true_negatives = 46
false_positives = 7
false_negatives = 13

# true predictions (pos/neg) / total observations
accuracy = (true_positives + true_negatives) / (true_positives + true_negatives + false_positives + false_negatives)

# true positives / (all actual positives)
recall = true_positives / (true_positives + false_negatives)

# true positives / (all actual positive predictions)
precision = true_positives / (true_positives + false_positives)




print("Cat-classifier (where 'cat' is the positive prediction)")
print("Accuracy:", accuracy)
print("Recall:", recall)
print("Precision:", precision)


Cat-classifier (where 'cat' is the positive prediction)
Accuracy: 0.8
Recall: 0.723404255319149
Precision: 0.8292682926829268


### 3. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.`

- Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found <a href='https://ds.codeup.com/data/c3.csv'> here.</a>

`Use the predictions dataset and pandas to help answer the following questions:`


    
    

- An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible.
    - Which evaluation metric would be appropriate here? 
    - Which model would be the best fit for this use case?

Possible outcomes :
defect present, predicted defect
defect present, predicted OK
no defects, predicted OK
no defects, predicted defect

**positive**: defect present
use recall for quality control

In [11]:
#acquire data 
# acquire the data
df = pd.read_csv("https://ds.codeup.com/data/c3.csv")
df.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

In [12]:
# assign the value of the most frequent 
df["baseline"] = df.actual.value_counts().index[0]
df.head()

actual     model1  model2     model3   baseline
0  No Defect  No Defect  Defect  No Defect  No Defect
1  No Defect  No Defect  Defect     Defect  No Defect
2  No Defect  No Defect  Defect  No Defect  No Defect
3  No Defect     Defect  Defect     Defect  No Defect
4  No Defect  No Defect  Defect  No Defect  No Defect

In [13]:
#recall for QC
tp = 8
tn = 182
fp = 2
fn = 8

recall = tp / (tp + fn)
recall

0.5

In [14]:
#POSITTIVE : DEFECT PRESENT 
positive = "Defect"

# Subset.actual == "Defect" (we've defined Defect as our positive)
subset = df[df.actual == positive]

# Trues are True Positives from Model1 predictions
# True Positive + False Negatives = Total number of actual positives

# Recall AKA True Positive Rate AKA Sensitity
# TP / (TP + FN)
# TP / (Actual positives)

print("# of TP", (subset.actual == subset.model1).sum()) # Summing all the trues gives us # of True Positives
# Number of True Positives is our numerator with Recall

# Denominor in recall = TP + FN = NUMBER of ACTUAL POSITIVES
print("# of TP + FN", subset.shape[0], " = total # of actual positives")

# of TP 8
# of TP + FN 16  = total # of actual positives


In [18]:
#confusion matrix for model # 1
confusion_matrix(df.actual, df.model1,
                labels =['No Defect', 'Defect'])

array([[182,   2],
       [  8,   8]])

In [15]:
# MODEL EVALUATION # 1
positive = "Defect"

# subset: occurences positive case was correct
subset = df[df.actual == positive]

model_recall = (subset.actual == subset.model1).mean()
baseline_recall = (subset.baseline == subset.actual).mean()

print("Model 1")
print(f"Model recall: {model_recall:.2%}")
print(f"Baseline recall: {baseline_recall:.2%}")

Model 1
Model recall: 50.00%
Baseline recall: 0.00%


In [16]:
#Model Evaluation # 2
positive = "Defect"

subset = df[ducks.actual == positive]
model_recall = (subset.actual == subset.model2).mean()
baseline_recall = (subset.baseline == subset.actual).mean()

print("Model 2")
print(f"Model recall: {model_recall:.2%}")
print(f"Baseline recall: {baseline_recall:.2%}")

Model 2
Model recall: 56.25%
Baseline recall: 0.00%


In [17]:
#Model Evaluation # 3.
positive = "Defect"

subset = df[df.actual == positive]
model_recall = (subset.actual == subset.model3).mean()
baseline_recall = (subset.baseline == subset.actual).mean()

print("Model 3")
print(f"Model recall: {model_recall:.2%}")
print(f"Baseline recall: {baseline_recall:.2%}")

Model 3
Model recall: 81.25%
Baseline recall: 0.00%


**Takeaways**
Quality Control should use Model 3 as it perfeorms the best with recall.

- Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. 
    - Which evaluation metric would be appropriate here? 
    - Which model would be the best fit for this use case?

In [18]:
#Trying to minimize false positives 
positive = "Defect"

In [19]:
#Model Evaluation # 1

subset = ducks[ducks.model1 == positive]

model_precision = (subset.actual == subset.model1).mean()

subset = ducks[ducks.baseline == positive]
baseline_precision = (subset.actual == subset.baseline).mean()

print("Model 1")
print(f"Model precision: {model_precision:.2%}")
print(f"Baseline precision: {baseline_precision:.2%}")

Model 1
Model precision: 80.00%
Baseline precision: nan%


In [21]:
# Model Evaluation # 2
subset = ducks[ducks.model2 == positive]
model_precision = (subset.actual == subset.model2).mean()
subset = ducks[ducks.baseline == positive]
baseline_precision = (subset.actual == subset.baseline).mean()

print("Model 2")
print(f"Model precision: {model_precision:.2%}")
print(f"Baseline precision: {baseline_precision:.2%}")

Model 2
Model precision: 10.00%
Baseline precision: nan%


In [22]:
# Model Evaluation # 3
subset = ducks[ducks.model3 == positive]
model_precision = (subset.actual == subset.model3).mean()
subset = ducks[ducks.baseline == positive]
baseline_precision = (subset.actual == subset.baseline).mean()

print("Model 3")
print(f"Model precision: {model_precision:.2%}")
print(f"Baseline precision: {baseline_precision:.2%}")

Model 3
Model precision: 13.13%
Baseline precision: nan%


**Takeaways** : Use Model 1 to minimize false positives
    

### 4 .  You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

 Several models have already been developed with the data, and you can find their results <a href='https://ds.codeup.com/data/gives_you_paws.csv'> here.</a>

In [21]:
# read data 
df = pd.read_csv("https://ds.codeup.com/data/gives_you_paws.csv")
df.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [22]:
df.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   actual  5000 non-null   object
 1   model1  5000 non-null   object
 2   model2  5000 non-null   object
 3   model3  5000 non-null   object
 4   model4  5000 non-null   object
dtypes: object(5)
memory usage: 195.4+ KB


In [27]:
df.describe()

actual model1 model2 model3 model4
count    5000   5000   5000   5000   5000
unique      2      2      2      2      2
top       dog    dog    cat    dog    dog
freq     3254   2937   3212   2508   4253

### 5. Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

- <a href = 'https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html'>sklearn.metrics.accuracy_score</a>
- <a href='https://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_score.html'>sklearn.metrics.precision_score</a>
- <a href='https://scikit-learn.org/stable/modules/generated/sklearn.metrics.recall_score.html'>sklearn.metrics.recall_score</a>
- <a href='https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html'>sklearn.metrics.classification_report</a>

In [29]:
df.shape

(5000, 5)

In [30]:
#value count
df.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

**Workflow**: 
  1. Create a baseline 
  2. Get all the model columns
  
Im using dogs as positive

- a. In terms of accuracy, how do the various models compare to the baseline model? 
    - Are any of the models better than the baseline?


In [25]:
#create a baseline 
df["baseline"] = df.actual.value_counts().index[0]
df.head()

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
4    cat    cat    cat    dog    dog      dog

In [26]:
# .loc[starting_row:ending_row, starting_column:ending_column]

models = df.loc[:, "model1":"baseline"].columns.tolist()
models

['model1', 'model2', 'model3', 'model4', 'baseline']

In [ ]:
#get accuracy in dictionaru from each model
output = model

In [19]:
# import sklearn functions
from sklearn.metrics import precision_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import classification_report

In [20]:
target_names = df.actual.unique()

In [43]:

cat_dog = classification_report(df.actual, df.model1, target_names=target_names, output_dict=True)
pd.DataFrame(cat_dog)

cat          dog  accuracy    macro avg  weighted avg
precision     0.689772     0.890024    0.8074     0.789898      0.820096
recall        0.815006     0.803319    0.8074     0.809162      0.807400
f1-score      0.747178     0.844452    0.8074     0.795815      0.810484
support    1746.000000  3254.000000    0.8074  5000.000000   5000.000000

In [57]:
# Model Evaluation # 1
print("Model 1")
print(classification_report(df.actual, 
                                   df.model1, 
                                   target_names=target_names, 
                                   ))

Model 1
              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000



In [64]:
print("Model 1 w/ ouput_dict")
pd.DataFrame(classification_report(df.actual, 
                                   df.model1, 
                                   target_names=target_names, 
                                   output_dict=True))

Model 1 w/ ouput_dict


cat          dog  accuracy    macro avg  weighted avg
precision     0.689772     0.890024    0.8074     0.789898      0.820096
recall        0.815006     0.803319    0.8074     0.809162      0.807400
f1-score      0.747178     0.844452    0.8074     0.795815      0.810484
support    1746.000000  3254.000000    0.8074  5000.000000   5000.000000

In [ ]:
#precesion model 1: 0.890024	

In [65]:
#Model Evaluation # 2
print('Model 2')
print(classification_report(df.actual,
                                  df.model2,
                                  target_names=target_names))

Model 2
              precision    recall  f1-score   support

         cat       0.48      0.89      0.63      1746
         dog       0.89      0.49      0.63      3254

    accuracy                           0.63      5000
   macro avg       0.69      0.69      0.63      5000
weighted avg       0.75      0.63      0.63      5000



In [66]:
print("Model 2 w/ ouput_dict")
pd.DataFrame(classification_report(df.actual, 
                                   df.model2, 
                                   target_names=target_names, 
                                   output_dict=True))

Model 2 w/ ouput_dict


cat          dog  accuracy    macro avg  weighted avg
precision     0.484122     0.893177    0.6304     0.688649      0.750335
recall        0.890607     0.490781    0.6304     0.690694      0.630400
f1-score      0.627269     0.633479    0.6304     0.630374      0.631310
support    1746.000000  3254.000000    0.6304  5000.000000   5000.000000

In [ ]:
#precesion model : 0.893177

In [51]:
#model Evaluation # 3
print('Model 3')
print(classification_report(df.actual,
                            df.model3,
                           target_names=target_names))

Model 3
              precision    recall  f1-score   support

         cat       0.36      0.51      0.42      1746
         dog       0.66      0.51      0.57      3254

    accuracy                           0.51      5000
   macro avg       0.51      0.51      0.50      5000
weighted avg       0.55      0.51      0.52      5000



In [58]:
#Model Evaluation #4 
print('Model 4')
print(classification_report(df.actual,
                           df.model4,
                           target_names=target_names))

Model 4
              precision    recall  f1-score   support

         cat       0.81      0.35      0.48      1746
         dog       0.73      0.96      0.83      3254

    accuracy                           0.74      5000
   macro avg       0.77      0.65      0.66      5000
weighted avg       0.76      0.74      0.71      5000



In [63]:
print("Model 4 w/ ouput_dict")
pd.DataFrame(classification_report(df.actual, 
                                   df.model4, 
                                   target_names=target_names, 
                                   output_dict=True))

Model 4 w/ ouput_dict


cat          dog  accuracy    macro avg  weighted avg
precision     0.807229     0.731249    0.7426     0.769239      0.757781
recall        0.345361     0.955747    0.7426     0.650554      0.742600
f1-score      0.483755     0.828560    0.7426     0.656157      0.708154
support    1746.000000  3254.000000    0.7426  5000.000000   5000.000000

In [59]:
df.head()

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
4    cat    cat    cat    dog    dog      dog

` Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:`




### **Dog as positive case**
* Phase I: Recall(optimizing for true positive cases) / (all positve cases)
    - Model 4 : Recall is 0.955
* Phase II: Precision (minimizing false positives: saying dog, but have a cat))
   - Model 1 : Precision is 0.890024

### **Cat as positive case**
* Phase I: Recall(optimizing for true positive cases) / (all positve cases)
    - Model 2 : Recall is 0.89
* Phase II: Precision (minimizing false positives: saying cat, but have a dog))
   - Model 1 : Precision is  0.81 

### Post-Walkthrough Solutions